src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +615,188 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
     private static <T> DispatchResult<T> emptyDispatchResult() {
         return (DispatchResult<T>) EMPTY_DISPATCH_RESULT;
     }
+
+    private class BatchObserverImpl implements BatchObserver<V> {
+        private final CompletableFuture<List<V>> valuesFuture;
+        private final List<K> keys;
+        private final List<Object> callContexts;
+        private final List<CompletableFuture<V>> queuedFutures;
+
+        private final List<K> clearCacheKeys = new ArrayList<>();
+        private final List<V> completedValues = new ArrayList<>();
+        private int idx = 0;
```

Comment:
in a reactive world, the `onNext` callback can happen at any time on any thread.   There is no happens before here for values inherently...  So these fields need to be synchronised in some manner.